community/bolt/src/main/java/org/neo4j/bolt/runtime/DefaultBoltConnection.java

```diff
@@ -188,6 +199,12 @@ protected boolean processNextBatch( int batchCount )
                     loop = machine.shouldStickOnThread();
                     waitForMessage = loop;
                 }
+
+                // we processed all pending messages, let's flush underlying channel
+                if ( queue.size() == 0 || maxBatchSize == 1 )
```

Comment:
Does this change the previous flush policy?
We used to flush after every message in `MessageProcessingHandler` and now will only flush after all queued messages have been processed. Is my understanding correct?

Reply:
Previously, `BoltMessagingProtocolHandlerImpl` was keeping track of the number of inbound messages received and number of completed ones ([here](https://github.com/neo4j/neo4j/blob/3.4/community/bolt/src/main/java/org/neo4j/bolt/v1/transport/BoltMessagingProtocolHandlerImpl.java#L118) and [here](https://github.com/neo4j/neo4j/blob/3.4/community/bolt/src/main/java/org/neo4j/bolt/v1/transport/BoltMessagingProtocolHandlerImpl.java#L130)). So, if  `RUN`, `PULLALL` sequence is received, the buffers were flushed only after `PULLALL` processing is finished (given that we don't fill up the internal buffers in `ChunkedOutput`).

We still stick to that behaviour, but now have more control on when to flush.